### PR TITLE
Added Input validation VR446

### PR DESCRIPTION
**Pull request by Mech0z**
## Description

For all Create and Updates Charge Name must not exceed 50 characters

## References

* #194

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using GreenEnergyHub.Charges.Domain.ChangeOfCharges.Transaction;
+
+namespace GreenEnergyHub.Charges.Application.Validation.InputValidation.ValidationRules
+{
+    public class ChargeNameLengthValidationRule : IValidationRule
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs:19]: Like @lasrinnil, I'd like name that express what we are actually asserting. E.g. `ChargeNameHasMinimumLength`.

Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs:19]: nit: Personally I'd like to remove the `Validation` part of all the rule names. But on the other hand, it's an inconsistency compared to our general naming convention...

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        private readonly ChargeCommand _chargeCommand;
+        private int _maximumChargeNameLength = 50;
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs:22]: const/readonly?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/ValidationRuleIdentifier.cs [hunk 3]
```diff
@@ -30,5 +30,6 @@ public enum ValidationRuleIdentifier
         InvalidResolutionTypeOnTariffOnCreate, // VR505-1
         InvalidResolutionTypeOnFeeOnCreate, // VR505-2
         InvalidResolutionTypeOnSubscriptionOnCreate, // VR505-3
+        MaximumChargeNameLengthExceeded, // VR446
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/ValidationRuleIdentifier.cs:33]: Oh, interesting naming thing. The rules seems to lean towards a naming implying the requirement that must be fulfilled, whereas the identifiers seem to be named according to the failure to meet the requirement. I think we should align the naming.

@lasrinnil, you might wanna contribute to this topic?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    public class ChargeNameLengthValidationRuleTests
+    {
+        [Theory]
+        [InlineAutoMoqData(49, true)]
+        [InlineAutoMoqData(50, true)]
+        [InlineAutoMoqData(51, false)]
+        public void ChargeNameLengthValidationRule_WhenCalledWithChargeNameLength_EqualsExpectedResult(int chargeNameLength, bool expected, [NotNull] ChargeCommand command)
+        {
+            command.ChargeOperation.ChargeName =
+                string.Join(string.Empty, Enumerable.Repeat(0, chargeNameLength).Select(n => "a"));
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs:35]: To much mental parsing

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+namespace GreenEnergyHub.Charges.Tests.Application.Validation.InputValidation.ValidationRules
+{
+    [UnitTest]
+    public class ChargeNameLengthValidationRuleTests
+    {
+        [Theory]
+        [InlineAutoMoqData(49, true)]
+        [InlineAutoMoqData(50, true)]
+        [InlineAutoMoqData(51, false)]
+        public void ChargeNameLengthValidationRule_WhenCalledWithChargeNameLength_EqualsExpectedResult(int chargeNameLength, bool expected, [NotNull] ChargeCommand command)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs:32]: 50 doesn't seem to be an arbitrary number. Perhaps it should be included in the test name somehow?

**Review by BjarkeMeier — APPROVED** [review]
Good job. Left a few comments. Perhaps you and @lasrinnil, could settle the naming things?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs (hunk 5 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs]: Changed in 2c8d63d60dc777e023dfbc66953aa421d83099a2

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs (hunk 4 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRuleTests.cs]: Moved to separate function in 2c8d63d60dc777e023dfbc66953aa421d83099a2

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs (hunk 2 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs]: Fixed in 2c8d63d60dc777e023dfbc66953aa421d83099a2

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs (hunk 1 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs]: Changed in 9e16ea5ca352f62a526b091ce5a8b37347dd25b2

Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/InputValidation/ValidationRules/ChargeNameLengthValidationRule.cs]: Changed in 9e16ea5ca352f62a526b091ce5a8b37347dd25b2

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/ValidationRuleIdentifier.cs (hunk 3 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/ValidationRuleIdentifier.cs]: Changed to ChargeNameHasMinimumLengthRule in 47830d8477942decadf908208f4058983ad25c28

**Review by lasrinnil — APPROVED** [review]

**Pull request merged**